File: commands/ascend.ts
import * as Discord from 'discord.js';
import { Command } from './classes/Command';
import { getMaxLevel } from '../utils/upgrades';
import { format } from '../utils/math';
import { getAscensionState, processAscend } from '../utils/ascend';

class Ascend extends Command {
  constructor(client: any) {
    super(client, 'ascend', 'Ascend to reset stuff but get more stuff', [], { blame: 'ei' });
  }

  async run(interaction: any): Promise<void> {
    const state = await getAscensionState(this.client, interaction.user.id);

    if (!state.canAscend) {
      await interaction.editReply({
        embeds: [new Discord.EmbedBuilder()
          .setColor('#FFA500')
          .setTitle('Cannot ascend')
          .setDescription(`You need at least 500 dinonuggies to ascend. You have ${format(state.dinonuggies)} dinonuggies.`)
          .setFooter({ text: 'so no one ascends and complains they cant buy anything from it' })],
      });
      return;
    }

    const confirmEmbed = new Discord.EmbedBuilder()
      .setColor('#FFA500')
      .setTitle('Ascension Confirmation')
      .setDescription(`Are you sure you want to ascend?
- Your dinonuggies (${format(state.dinonuggies)}) will be converted to ${format(state.dinonuggies)} heavenly nuggies which can be used to buy better upgrades
- All your upgrades, credits, bitcoins, dinonuggies, and streak will reset
- If you ascend with all upgrades maxed, you will gain an ascension level, which increases the level cap by 10

Your current upgrades:
• Multiplier amount: ${state.multiplierAmountLevel}/${state.currentMaxLevel}
• Multiplier rarity: ${state.multiplierRarityLevel}/${state.currentMaxLevel}
• Beki cooldown: ${state.bekiLevel}/${state.currentMaxLevel}

${state.allMaxed ? `Your ascension level will increase from ${state.ascensionLevel} to ${state.ascensionLevel + 1} if you ascend now, allowing you to buy upgrades up to level ${getMaxLevel(state.ascensionLevel + 1)}` : `Your ascension level will remain at ${state.ascensionLevel} as not all upgrades are maxed.`}`)
      .setFooter({ text: 'what even is this game now' });

    const row = new Discord.ActionRowBuilder()
      .addComponents(
        new Discord.ButtonBuilder()
          .setCustomId('confirm_ascend')
          .setLabel('Confirm')
          .setStyle(Discord.ButtonStyle.Success),
        new Discord.ButtonBuilder()
          .setCustomId('cancel_ascend')
          .setLabel('Cancel')
          .setStyle(Discord.ButtonStyle.Danger),
      );

    const confirmMessage = await interaction.editReply({
      embeds: [confirmEmbed],
      components: [row],
    });

    const filter = (i: any) => ['confirm_ascend', 'cancel_ascend'].includes(i.customId) && i.user.id === interaction.user.id;
    const collector = confirmMessage.createMessageComponentCollector({ filter, time: 30000 });

    let clicked = false;
    collector.on('collect', async (i: any) => {
      clicked = true;
      await confirmMessage.edit({ components: [] });

      if (i.customId === 'confirm_ascend') {
        const result = await processAscend(this.client, interaction.user.id);

        if (result.status === 'too_few') {
          await interaction.followUp({
            embeds: [new Discord.EmbedBuilder()
              .setColor('#FFA500')
              .setTitle('Cannot ascend')
              .setDescription(`You need at least 500 dinonuggies to ascend. You have ${format(result.dinonuggies)} dinonuggies.`)
              .setFooter({ text: 'so no one ascends and complains they cant buy anything from it' })],
          });
          return;
        }

        const resultEmbed = new Discord.EmbedBuilder()
          .setColor('#00AA00')
          .setTitle('Ascension Successful!')
          .setDescription(`You have ascended!
- Gained ${format(result.gained)} heavenly nuggies
- All other stuff reset
${result.allMaxed ? `- Ascension level increased to ${result.ascensionLevel}` : '- Ascension level remained the same'}
- New max level: ${result.newMaxLevel}`)
          .setFooter({ text: 'what even is this game now' });

        await interaction.followUp({ embeds: [resultEmbed] });
      } else {
        await interaction.followUp({ content: 'Ascension cancelled.', embeds: [] });
      }
    });

    collector.on('end', async () => {
      if (!clicked) {
        await interaction.followUp({ content: 'Ascension timed out', embeds: [] });
      }
    });
  }
}

export default Ascend;

File: src/commands/ascend.rs
use crate::{Context, Error, utils, upgrades};
use poise::serenity_prelude as serenity;
use futures::StreamExt;

/// Ascend to reset stuff but get more stuff
#[poise::command(slash_command)]
pub async fn ascend(ctx: Context<'_>) -> Result<(), Error> {
    ctx.defer().await?;
    let user_id = ctx.author().id.to_string();
    let user = ctx.data().db.get_user(&user_id).await?;
    
    let current_max = upgrades::get_max_level(user.ascension_level);
    let all_maxed = user.multiplier_amount_level >= current_max && 
                    user.multiplier_rarity_level >= current_max && 
                    user.beki_level >= current_max;

    if user.dinonuggies < 500.0 {
        ctx.send(poise::CreateReply::default()
            .embed(serenity::CreateEmbed::new()
                .color(0xFFA500)
                .title("Cannot ascend")
                .description(format!("You need at least 500 dinonuggies to ascend. You have {} dinonuggies.", utils::format(user.dinonuggies)))
                .footer(serenity::CreateEmbedFooter::new("so no one ascends and complains they cant buy anything from it")))
        ).await?;
        return Ok(());
    }

    let next_max = upgrades::get_max_level(if all_maxed { user.ascension_level + 1 } else { user.ascension_level });

    let desc = format!(
        "Are you sure you want to ascend?\n\
        - Your dinonuggies ({}) will be converted to {} heavenly nuggies which can be used to buy better upgrades\n\
        - All your upgrades, credits, bitcoins, dinonuggies, and streak will reset\n\
        - If you ascend with all upgrades maxed, you will gain an ascension level, which increases the level cap by 10\n\n\
        Your current upgrades:\n\
        • Multiplier amount: {}/{}\n\
        • Multiplier rarity: {}/{}\n\
        • Beki cooldown: {}/{}\n\n\
        {}",
        utils::format(user.dinonuggies),
        utils::format(user.dinonuggies),
        user.multiplier_amount_level, current_max,
        user.multiplier_rarity_level, current_max,
        user.beki_level, current_max,
        if all_maxed {
            format!("Your ascension level will increase from {} to {} if you ascend now, allowing you to buy upgrades up to level {}", user.ascension_level, user.ascension_level + 1, next_max)
        } else {
            format!("Your ascension level will remain at {} as not all upgrades are maxed.", user.ascension_level)
        }
    );

    let ctx_id = ctx.id();
    let confirm_id = format!("{}confirm", ctx_id);
    let cancel_id = format!("{}cancel", ctx_id);

    let reply = ctx.send(poise::CreateReply::default()
        .embed(serenity::CreateEmbed::new()
            .color(0xFFA500)
            .title("Ascension Confirmation")
            .description(desc)
            .footer(serenity::CreateEmbedFooter::new("what even is this game now")))
        .components(vec![serenity::CreateActionRow::Buttons(vec![
            serenity::CreateButton::new(&confirm_id).label("Confirm").style(serenity::ButtonStyle::Success),
            serenity::CreateButton::new(&cancel_id).label("Cancel").style(serenity::ButtonStyle::Danger),
        ])])
    ).await?;

    let mut interaction_stream = reply
        .message()
        .await?
        .await_component_interactions(ctx.serenity_context())
        .timeout(std::time::Duration::from_secs(30))
        .stream();

    if let Some(mci) = interaction_stream.next().await {
        if mci.user.id != ctx.author().id {
            mci.create_response(ctx.serenity_context(), serenity::CreateInteractionResponse::Message(
                serenity::CreateInteractionResponseMessage::new().content("Not for you!").ephemeral(true)
            )).await?;
            return Ok(());
        }

        if mci.data.custom_id == confirm_id {
            let gained = user.dinonuggies;
            ctx.data().db.ascend_user(&user_id, all_maxed).await?;

            mci.create_response(ctx.serenity_context(), serenity::CreateInteractionResponse::UpdateMessage(
                serenity::CreateInteractionResponseMessage::new()
                    .embed(serenity::CreateEmbed::new()
                        .color(0x00AA00)
                        .title("Ascension Successful!")
                        .description(format!(
                            "You have ascended!\n- Gained {} heavenly nuggies\n- All other stuff reset\n{}\n- New max level: {}",
                            utils::format(gained),
                            if all_maxed { format!("- Ascension level increased to {}", user.ascension_level + 1) } else { "- Ascension level remained the same".to_string() },
                            next_max
                        ))
                        .footer(serenity::CreateEmbedFooter::new("what even is this game now")))
                    .components(vec![])
            )).await?;
        } else {
            mci.create_response(ctx.serenity_context(), serenity::CreateInteractionResponse::UpdateMessage(
                serenity::CreateInteractionResponseMessage::new()
                    .content("Ascension cancelled.")
                    .embeds(vec![])
                    .components(vec![])
            )).await?;
        }
    } else {
        ctx.send(poise::CreateReply::default()
            .content("Ascension timed out.")
            .components(vec![])
        ).await?;
    }

    Ok(())
}
